File: app/GOExercise/GOExercise/Services/Analytics.swift
import Foundation

#if canImport(TelemetryDeck)
import TelemetryDeck
#endif

/// 計測する行動イベント。離脱・課金ファネルを後から分析できるよう、
/// チェックリストの主要ステップに対応させている。
enum AnalyticsEvent {
    case appOpen
    case onboardingCompleted
    case recordCreated(category: String)
    case paywallViewed(product: String)
    case purchaseStarted(product: String)
    case purchaseCompleted(product: String)
    case dataExported
    case dataDeleted

    /// TelemetryDeck の signal 名 (英小文字スネーク)。
    var name: String {
        switch self {
        case .appOpen:             return "app_open"
        case .onboardingCompleted: return "onboarding_complete"
        case .recordCreated:       return "record_created"
        case .paywallViewed:       return "view_paywall"
        case .purchaseStarted:     return "start_purchase"
        case .purchaseCompleted:   return "purchase_complete"
        case .dataExported:        return "data_exported"
        case .dataDeleted:         return "data_deleted"
        }
    }

    /// 個人を特定しない補助パラメータのみ。
    var parameters: [String: String] {
        switch self {
        case let .recordCreated(category):     return ["category": category]
        case let .paywallViewed(product):      return ["product": product]
        case let .purchaseStarted(product):    return ["product": product]
        case let .purchaseCompleted(product):  return ["product": product]
        default:                               return [:]
        }
    }
}

/// 計測の抽象。アプリ本体はこの protocol 越しに track するだけで、
/// 実体 (TelemetryDeck / Noop) を意識しない。
protocol AnalyticsService: Sendable {
    func track(_ event: AnalyticsEvent)
}

/// 何も送らない実装。テスト・DEBUG・App ID 未設定時のデフォルト。
struct NoopAnalytics: AnalyticsService {
    func track(_ event: AnalyticsEvent) {}
}

#if canImport(TelemetryDeck)
/// TelemetryDeck (プライバシー配慮型) への送信実装。
struct TelemetryDeckAnalytics: AnalyticsService {
    func track(_ event: AnalyticsEvent) {
        TelemetryDeck.signal(event.name, parameters: event.parameters)
    }
}
#endif

/// アプリ全体から呼ぶ計測ファサード。
///
/// 既定は `NoopAnalytics` なので **App ID を設定するまで一切データを送らない**。
/// これにより「Apple Developer 加入 + TelemetryDeck App ID 設定」までは現状の
/// ゼロ収集 (プライバシーラベル「データ収集なし」) を維持できる。
@MainActor
enum Analytics {
    /// 差し替え可能な実体 (テストでは Noop のまま)。
    static var service: AnalyticsService = NoopAnalytics()

    /// 匿名分析を共有するか (既定 true = 匿名 ON)。設定の「利用状況の分析を共有」でオプトアウト可。
    /// UserDefaults に永続。false の間は [track]/[configureIfPossible] が一切送信・初期化しない。
    /// (Android `Analytics.consentGranted` と対称)。
    static let analyticsEnabledKey = "analyticsEnabled"
    static var isEnabled: Bool {
        get { UserDefaults.standard.object(forKey: analyticsEnabledKey) as? Bool ?? true }
        set { UserDefaults.standard.set(newValue, forKey: analyticsEnabledKey) }
    }

    static func track(_ event: AnalyticsEvent) {
        guard isEnabled else { return } // オプトアウト中は送らない
        service.track(event)
    }

    /// 設定トグルからの opt-in/out 切替。OFF にしたら **その場で** SDK 実体を Noop に戻し、
    /// セッション中も完全に送信を止める(GPT-5.5/Claude 監査: 旧実装は track の gate のみで、
    /// 既に初期化済みの TelemetryDeck はセッション中 live のまま=厳格なレビュアが残留を問題視し得る)。
    /// ON は次回 configureIfPossible(or 即時再構成)で再有効化。
    static func setEnabled(_ enabled: Bool) {
        isEnabled = enabled
        if enabled {
            configureIfPossible()
        } else {
            service = NoopAnalytics()
        }
    }

    /// 起動時に一度だけ呼ぶ。**ユーザーが分析を許可** かつ App ID 設定済み かつ Release ビルドの
    /// ときだけ TelemetryDeck を有効化する。オプトアウト中は SDK 自体を起動しない。
    static func configureIfPossible() {
        #if canImport(TelemetryDeck) && !DEBUG
        guard isEnabled else { return } // オプトアウト中は初期化もしない
        guard let appID = telemetryAppID, !appID.isEmpty else { return }
        TelemetryDeck.initialize(config: TelemetryDeck.Config(appID: appID))
        service = TelemetryDeckAnalytics()
        #endif
    }

    private static var telemetryAppID: String? {
        Bundle.main.object(forInfoDictionaryKey: "TelemetryDeckAppID") as? String
    }
}

File: app-android/app/src/main/java/com/goexercise/app/analytics/Analytics.kt
package com.goexercise.app.analytics

import android.content.Context
import android.util.Log
import com.goexercise.app.BuildConfig
import com.telemetrydeck.sdk.TelemetryDeck

/**
 * 計測する行動イベント。離脱・課金ファネルを後から分析できるよう、主要ステップに対応させる。
 * iOS `AnalyticsEvent`(Analytics.swift)と signal 名・パラメータを揃える。
 */
sealed interface AnalyticsEvent {
    /** TelemetryDeck の signal 名(英小文字スネーク)。iOS と一致させること。 */
    val name: String

    /** 個人を特定しない補助パラメータのみ。 */
    val parameters: Map<String, String> get() = emptyMap()

    data object AppOpen : AnalyticsEvent {
        override val name = "app_open"
    }

    data object OnboardingCompleted : AnalyticsEvent {
        override val name = "onboarding_complete"
    }

    data class RecordCreated(val category: String) : AnalyticsEvent {
        override val name = "record_created"
        override val parameters = mapOf("category" to category)
    }

    data class PaywallViewed(val product: String) : AnalyticsEvent {
        override val name = "view_paywall"
        override val parameters = mapOf("product" to product)
    }

    data class PurchaseStarted(val product: String) : AnalyticsEvent {
        override val name = "start_purchase"
        override val parameters = mapOf("product" to product)
    }

    data class PurchaseCompleted(val product: String) : AnalyticsEvent {
        override val name = "purchase_complete"
        override val parameters = mapOf("product" to product)
    }

    data object DataExported : AnalyticsEvent {
        override val name = "data_exported"
    }

    data object DataDeleted : AnalyticsEvent {
        override val name = "data_deleted"
    }
}

/**
 * 計測の抽象。アプリ本体はこの interface 越しに track するだけで、実体
 * (TelemetryDeck / Noop)を意識しない。iOS `AnalyticsService` protocol に対応。
 */
interface AnalyticsService {
    fun track(event: AnalyticsEvent)
}

/** 何も送らない実装。テスト・DEBUG・App ID 未設定時のデフォルト。 */
object NoopAnalytics : AnalyticsService {
    override fun track(event: AnalyticsEvent) {}
}

/** TelemetryDeck(プライバシー配慮型・匿名)への送信実装。 */
internal class TelemetryDeckAnalytics : AnalyticsService {
    override fun track(event: AnalyticsEvent) {
        TelemetryDeck.signal(event.name, params = event.parameters)
    }
}

/**
 * アプリ全体から呼ぶ計測ファサード。iOS `enum Analytics` に対応(グローバルな差し替え可能ファサード)。
 *
 * 既定は [NoopAnalytics] なので **App ID を設定するまで一切データを送らない**。これにより
 * 「TelemetryDeck App ID 設定」までは現状のゼロ収集(プライバシーラベル「データ収集なし」)を維持する。
 */
object Analytics {
    private const val TAG = "Analytics"

    /** 差し替え可能な実体(テストでは Noop のまま)。 */
    @Volatile
    var service: AnalyticsService = NoopAnalytics

    fun track(event: AnalyticsEvent) {
        service.track(event)
    }

    /**
     * 起動時に一度だけ呼ぶ。App ID が設定済み かつ Release ビルドのときだけ TelemetryDeck を有効化する。
     * iOS の `configureIfPossible()`(`#if !DEBUG` + App ID guard)と対称。
     * App ID が不正でも(`UUID.fromString` 例外)アプリは落とさない。
     */
    fun configureIfPossible(context: Context) {
        if (BuildConfig.DEBUG) return
        val appId = BuildConfig.TELEMETRYDECK_APP_ID
        if (appId.isBlank()) return
        runCatching {
            TelemetryDeck.start(context.applicationContext, TelemetryDeck.Builder().appID(appId))
            service = TelemetryDeckAnalytics()
        }.onFailure { Log.w(TAG, "TelemetryDeck init skipped: ${it.message}") }
    }
}
